torch/nn/utils/clip_grad.py

```diff
@@ -27,11 +27,7 @@ def clip_grad_norm_(parameters, max_norm, norm_type=2):
     if norm_type == inf:
         total_norm = max(p.grad.data.abs().max() for p in parameters)
     else:
-        total_norm = 0
-        for p in parameters:
-            param_norm = p.grad.data.norm(norm_type)
-            total_norm += param_norm.item() ** norm_type
-        total_norm = total_norm ** (1. / norm_type)
+        total_norm = torch.norm(torch.stack([torch.norm(p.grad.data, norm_type) for p in parameters]), norm_type)
```

Comment:
wait, norm of norms of segments is not generally equivalent with norm, right?

Reply:
I stopped there for a moment too, but turns out that it is.

Reply:
I though the same, but for 2 norm it works.
For two segments a and b: `(a.pow(2).sum().sqrt().pow(2) + b.pow(2).sum().sqrt().pow(2)).sqrt() = (a.pow(2).sum() + b.pow(2).sum()).sqrt()`.
Wouldn't work for squared norm though.

Reply:
Hmm don't you need a `pow` here for this to work though?

Reply:
The second norm is doing it.

Reply:
ohhh right. thanks for the explanation!

Reply:
I checked the math a few times to be sure that this is correct. In fact, this motivated one of my first suggestions for improving the performance, as we potentially spend time taking a root and then raising it to the same power. (That suggestion didn't work due to memory issues).

If it helps, [this script](https://gist.github.com/Enealor/b52e0cf7a67803ccff006f8e6dd6ada9) prints the result of computing the norm in each way. You can experimentally verify it works correctly for other powers quickly this way.